lib/jsonapi/utils/exceptions.rb

```diff
@@ -4,20 +4,57 @@ module JSONAPI
   module Utils
     module Exceptions
       class ActiveRecord < ::JSONAPI::Exceptions::Error
-        attr_accessor :object
+        attr_accessor :object, :resource, :relationships, :relationship_keys, :foreign_keys
 
-        def initialize(object)
+        def initialize(object, request, context)
           @object = object
+          request = request
+          resource_klass = request.resource_klass
+          @resource = resource_klass.new(object, context)
+
+          # Need to reflect on object's associations for error reporting.
+          @relationships     = resource_klass._relationships.values
+          @relationship_keys = @relationships.map(&:name).map(&:to_sym)
+          @foreign_keys      = @relationships.map(&:foreign_key).map(&:to_sym)
         end
 
         def errors
```

Comment:
Also a good suggestion. I'll look at this more in-depth this evening after I clear my head a little.